### ISSUE TITLE
$default is not defined (:esm)

### ISSUE THREAD

**Issue by dehli** [thheller/shadow-cljs#1106]
Creating an issue from my comment in https://github.com/thheller/shadow-cljs/issues/1102#issuecomment-1490176887 since it seems to be a bug even when not using the new `:force-imports` property. 

When creating an `:esm` release build with `2.22.9`, there is a bug where you'll get:

```
Uncaught ReferenceError: $default is not defined
```

In the generated code, the only reference to `$default` is here:

`return Sf(new R(null,2,5,T,[w,new R(null,2,5,T,[$default,d],null)],null)`

**Comment by thheller** [comment]
I don't see how my previous change leads to this. Can you make the release build with `--pseudo-names` to find out the context of that? I can't reproduce it and how little to go on.

**Comment by dehli** [comment]
Sure! Sorry for not providing more info. The only thing I changed is upgrading from `2.22.8` to `2.22.9` (and adding `--pseudo-names`).

The offending line seems to be related to cljs internals.

```cljs
  return $cljs$core$merge$$.$cljs$core$IFn$_invoke$arity$variadic$($cljs$core$prim_seq$cljs$0core$0IFn$0_invoke$0arity$02$$([$cljs$core$into$$.$cljs$core$IFn$_invoke$arity$2$($cljs$core$PersistentArrayMap$EMPTY$$, $cljs$core$reverse$$(function() {
    return function $cljs$pprint$map_params_$_iter__84944$$($s__84945$$) {
      return new $cljs$core$LazySeq$$(null, function() {
        for (;;) {
          var $s__84945__$2_temp__5804__auto__$jscomp$253$$ = $cljs$core$seq$$($s__84945$$);
          if ($s__84945__$2_temp__5804__auto__$jscomp$253$$) {
            if ($cljs$core$chunked_seq_QMARK_$$($s__84945__$2_temp__5804__auto__$jscomp$253$$)) {
              var $JSCompiler_inline_result$jscomp$824_c__5521__auto__$jscomp$112_vec__84954_vec__84957$$ = $cljs$core$_chunked_first$$($s__84945__$2_temp__5804__auto__$jscomp$253$$), $size__5522__auto__$jscomp$112$$ = $cljs$core$count$$($JSCompiler_inline_result$jscomp$824_c__5521__auto__$jscomp$112_vec__84954_vec__84957$$), $b__84947_name$jscomp$158$$ = $cljs$core$chunk_buffer$$($size__5522__auto__$jscomp$112$$);
              a: {
                for (var $i__84946$jscomp$inline_3564$$ = 0;;) {
                  if ($i__84946$jscomp$inline_3564$$ < $size__5522__auto__$jscomp$112$$) {
                    var $default$$jscomp$inline_3568_vec__84948$jscomp$inline_3565_vec__84951$jscomp$inline_3567$$ = $cljs$core$_nth$$($JSCompiler_inline_result$jscomp$824_c__5521__auto__$jscomp$112_vec__84954_vec__84957$$, $i__84946$jscomp$inline_3564$$), $name$jscomp$inline_3566$$ = $cljs$core$nth$cljs$0core$0IFn$0_invoke$0arity$03$$($default$$jscomp$inline_3568_vec__84948$jscomp$inline_3565_vec__84951$jscomp$inline_3567$$, 0, null);
                    $default$$jscomp$inline_3568_vec__84948$jscomp$inline_3565_vec__84951$jscomp$inline_3567$$ = $cljs$core$nth$cljs$0core$0IFn$0_invoke$0arity$03$$($default$$jscomp$inline_3568_vec__84948$jscomp$inline_3565_vec__84951$jscomp$inline_3567$$, 1, null);
                    $default$$jscomp$inline_3568_vec__84948$jscomp$inline_3565_vec__84951$jscomp$inline_3567$$ = $cljs$core$nth$cljs$0core$0IFn$0_invoke$0arity$03$$($default$$jscomp$inline_3568_vec__84948$jscomp$inline_3565_vec__84951$jscomp$inline_3567$$, 0, null);
                    $b__84947_name$jscomp$158$$.add(new $cljs$core$PersistentVector$$(null, 2, 5, $cljs$core$PersistentVector$EMPTY_NODE$$, [$name$jscomp$inline_3566$$, new $cljs$core$PersistentVector$$(null, 2, 5, $cljs$core$PersistentVector$EMPTY_NODE$$, [$default$$jscomp$inline_3568_vec__84948$jscomp$inline_3565_vec__84951$jscomp$inline_3567$$, $offset$jscomp$95$$], null)], null));
                    $i__84946$jscomp$inline_3564$$ += 1;
                  } else {
                    $JSCompiler_inline_result$jscomp$824_c__5521__auto__$jscomp$112_vec__84954_vec__84957$$ = !0;
                    break a;
                  }
                }
              }
              return $JSCompiler_inline_result$jscomp$824_c__5521__auto__$jscomp$112_vec__84954_vec__84957$$ ? $cljs$core$chunk_cons$$($cljs$core$chunk$$($b__84947_name$jscomp$158$$), $cljs$pprint$map_params_$_iter__84944$$($cljs$core$_chunked_rest$$($s__84945__$2_temp__5804__auto__$jscomp$253$$))) : $cljs$core$chunk_cons$$($cljs$core$chunk$$($b__84947_name$jscomp$158$$), null);
            }
            $JSCompiler_inline_result$jscomp$824_c__5521__auto__$jscomp$112_vec__84954_vec__84957$$ = $cljs$core$first$$($s__84945__$2_temp__5804__auto__$jscomp$253$$);
            $b__84947_name$jscomp$158$$ = $cljs$core$nth$cljs$0core$0IFn$0_invoke$0arity$03$$($JSCompiler_inline_result$jscomp$824_c__5521__auto__$jscomp$112_vec__84954_vec__84957$$, 0, null);
            $JSCompiler_inline_result$jscomp$824_c__5521__auto__$jscomp$112_vec__84954_vec__84957$$ = $cljs$core$nth$cljs$0core$0IFn$0_invoke$0arity$03$$($JSCompiler_inline_result$jscomp$824_c__5521__auto__$jscomp$112_vec__84954_vec__84957$$, 1, null);
            $cljs$core$nth$cljs$0core$0IFn$0_invoke$0arity$03$$($JSCompiler_inline_result$jscomp$824_c__5521__auto__$jscomp$112_vec__84954_vec__84957$$, 0, null);

            // This $default is undefined --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------->
            return $cljs$core$cons$$(new $cljs$core$PersistentVector$$(null, 2, 5, $cljs$core$PersistentVector$EMPTY_NODE$$, [$b__84947_name$jscomp$158$$, new $cljs$core$PersistentVector$$(null, 2, 5, $cljs$core$PersistentVector$EMPTY_NODE$$, [$default, $offset$jscomp$95$$], null)], null), $cljs$pprint$map_params_$_iter__84944$$($cljs$core$rest$$($s__84945__$2_temp__5804__auto__$jscomp$253$$)));
          }
          return null;
        }
      }, null, null);
    }($cljs$cst$727$params$$.$cljs$core$IFn$_invoke$arity$1$($def$jscomp$4$$));
```

I'll work on getting a repro for you.

**Comment by thheller** [comment]
Hmm yeah no clue what that is.

**Comment by dehli** [comment]
Here's a quick repro I put together. Let me know if I can provide anything else.

**Comment by thheller** [comment]
After changing `index.html` to properly load the ESM module `<script type="module" src="/out/main.js"></script>` I get no errors? Output code also doesn't contain `$default`? I just get the expected `"Hello woirld"` console log.

**Comment by dehli** [comment]
:man_facepalming: My browser was loading cached code. Will try to actually make it reproducible.

**Comment by thheller** [comment]
Maybe should use `:dev-http` so this doesn't happen. ;) It sets the proper headers so caches are always verified.

**Comment by dehli** [comment]
Still looking into it (searching through the project to see what's causing the compilation error), but I do see the following warning when I build which matches the generated code up above.

```
------ WARNING #1 -  -----------------------------------------------------------
 File: ~/.m2/repository/org/clojure/clojurescript/1.11.60/clojurescript-1.11.60.jar!/cljs/pprint.cljs:2541:19
--------------------------------------------------------------------------------
2538 | 
2539 |   (merge                                ; create the result map
2540 |     (into (array-map) ; start with the default values, make sure the order is right
2541 |           (reverse (for [[name [default]] (:params def)] [name [default offset]])))
-------------------------^------------------------------------------------------
 variable $default is undeclared
--------------------------------------------------------------------------------
2542 |     (reduce #(apply assoc %1 %2) {} (filter #(first (nth % 1)) (zipmap (keys (:params def)) params))) ; add the specified parameters, filtering out nils
2543 |     flags)); and finally add the flags
2544 | 
2545 | (defn- compile-directive [s offset]
--------------------------------------------------------------------------------
```

**Comment by thheller** [comment]
`cljs.pprint` compiles without warning for me. `default` there is also munged to `default$`? Not sure where `$default` is coming from? Do you load other code that messes with the cljs.compiler in some ways?

**Comment by dehli** [comment]
So what it ended up being was something off in my `.shadow-cljs` folder. I kept removing namespaces & deps and ended up with just `(:require [cljs.pprint])` and the error still showing (which wasn't reproducible in my other project). 

After I removed the `.shadow-cljs` folder, and rebuilt with 2.22.9, the `cljs.pprint` warning was gone.

No idea why it built fine on 2.22.8, but I'll close this issue and remember to clear that folder in the future if I see odd behavior. Thanks for your help!
